deafrica/platform/warn_unused_sandbox_volumes.py

```diff
@@ -0,0 +1,293 @@
+    else:
+        log.info(f"{days_to_delete} day warning email successfully sent")
+
+
+def warn_unused_sandbox_volumes(cluster_name, cron_schedule, dryrun):
+
+    log.info(f"dryrun : {dryrun}")
+    # configure boto3 clients
+    ec2_resource = boto3.resource("ec2")
+    ct_client = boto3.client("cloudtrail")
+    ses_client = boto3.client("ses")
+
+    k8s_namespace = "sandbox"
+    daysback = 90
+
+    time_now = datetime.now()
+    time_back = time_now - timedelta(days=daysback)
+    filters = filters = [
+        {
+            "Name": "tag:kubernetes.io/created-for/pvc/namespace",
+            "Values": [
+                k8s_namespace,
+            ],
+        },
+        {
+            "Name": "tag:kubernetes.io/created-for/pvc/name",
+            "Values": [
+                "*",
+            ],
+        },
+    ]
+
+    # get all users from cogneto
+    os.environ["AWS_DEFAULT_REGION"] = "us-west-2"
+    cognito_client = boto3.client("cognito-idp")
+    all_cognito_users = get_all_cognito_users(cognito_client, cluster_name)
+
+    volume_warnings = []  # collect some information about the volumes
+    count = 0
+
+    for volume in ec2_resource.volumes.filter(Filters=filters):
+        count += 1
+        response = ct_client.lookup_events(
+            LookupAttributes=[
+                {"AttributeKey": "ResourceName", "AttributeValue": volume.id},
+            ],
+            MaxResults=100,
+            StartTime=time_back,
+            EndTime=time_now,
+        )
+
+        attach_events = [
+            event
+            for event in response["Events"]
+            if event.get("EventName") in ["AttachVolume", "DetachVolume"]
+        ]
+
+        if len(attach_events) == 0 and volume.state == "available":
+            # no attachments in last 90 days and ebs is not in use
+            volume_last_attach_detach = None
+            volume_last_attach_detach_delta = None
+        if len(attach_events) > 0:
+            # attach events in the last 90 days
+            volume_last_attach_detach = attach_events[0]["EventTime"]
+            volume_last_attach_detach_delta = (
+                datetime.now(volume_last_attach_detach.tzinfo)
+                - volume_last_attach_detach
+            )
+
+        # get the persistant volume (pv) and persistant volume claim (pvc)
+        pv_name, pvc_name = get_user_claim(volume)
+        # get the email from the pvc
+        pvc_email = pvc_name[6:].replace("-40", "@").replace("-2e", ".")
+        # find the user for email
+        user, user_email = find_cognito_user(all_cognito_users, pvc_email)
+
+        # get number of days to delete (i.e. when volume is 90 days unused)
+        time_to_delete = (
+            timedelta(days=90) - volume_last_attach_detach_delta
+            if volume_last_attach_detach is not None
+            else timedelta(days=0)
+        )
+
+        # note, the deletion script is currently run monthly so the volumes
```

Comment:
might be worth rephrasing this so it doesn't immediately get out of date as soon as the workflow is updated to daily

perhaps a warning comment, "if it is run X"